be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -43,65 +44,60 @@ object GetMessagesByIdResponseHandler extends AskPatternConstants {
+      validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed]
   )(implicit system: ActorSystem): Boolean = {
-    if (receivedResponse.isEmpty || remainingAttempts <= 0) {
+    if (receivedResponse.isEmpty || receivedResponse.forall(elem => elem._2 <= 0)) {
       return receivedResponse.isEmpty
     }
 
-    var failedMessages: Map[Channel, Set[GraphMessage]] = Map.empty
+    var failedMessages: List[(GraphMessage, Int)] = Nil
     receivedResponse.foreach {
-      case (channel, messagesSet) =>
-        val failedMsgErrorMap = messagesThroughPipeline(validatorFlow, messagesSet)
-        if (failedMsgErrorMap.nonEmpty) {
-          failedMessages += channel -> failedMsgErrorMap.keySet
-
-          // only log a during last attempt
-          if (remainingAttempts == 1) {
-            println("Errors: " + failedMsgErrorMap.map {
-              case (msg, error) => error.toString + "\n On:\n" + prettyPrinter(msg)
-            })
+      case (graphMessage, retry) =>
+        if (retry > 0) {
+          messagesThroughPipeline(validatorFlow, graphMessage) match {
+            case Left(err) =>
+              failedMessages = (graphMessage, retry - 1) :: failedMessages
+              // only log a during last attempt
+              if (retry == 1) {
+                println("Errors: " + err.toString + "\n On:\n" + prettyPrinter(graphMessage))
+              }
+            case _ => /* DO NOTHING */
           }
         }
     }
 
-    passThroughPipeline(failedMessages, validatorFlow, remainingAttempts - 1)
+    passThroughPipeline(failedMessages.reverse, validatorFlow)
   }
 
-  // Push a set of message corresponding to one channel through the pipeline
+  // Push a message corresponding through the pipeline
   private def messagesThroughPipeline(
       validator: Flow[GraphMessage, GraphMessage, NotUsed],
-      messageSet: Set[GraphMessage]
-  )(implicit system: ActorSystem): Map[GraphMessage, PipelineError] = {
-    var failedGraphMessages: Map[GraphMessage, PipelineError] = Map.empty
-    messageSet.foreach { message =>
-      val singleRun = Source.single(message).via(validator).runWith(Sink.foreach {
-        case Left(err: PipelineError) =>
-          failedGraphMessages += message -> err
-
-        case _ => /* Nothing to do */
-      })
-      Await.ready(singleRun, duration)
+      message: GraphMessage
+  )(implicit system: ActorSystem): GraphMessage = {
+    val output = Source.single(message).via(validator).runWith(Sink.head)
+    Await.ready(output, duration).value.get match {
+      case Success(res @ _) => res
+      case err @ _          => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, err.toString, None))
     }
-
-    failedGraphMessages
   }
 
-  private def wrapMsgInPublish(map: Map[Channel, Set[Message]]): Map[Channel, Set[GraphMessage]] = {
-    map.map { case (channel, set) =>
-      channel -> set.map(msg =>
-        Right(JsonRpcRequest(
+  // Must return a list sorted by channel length
```

Comment:
Try to be more precise in your comments: a list of what?

Reply:
I added proper documentation